### PR TITLE
feat: add create table procedure

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

1. Introduce DDL Manager
2. Implement a create table procedure

It will make PR very large if I remove the router's Create method 🥲

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#1727

### CODE REVIEW

**Comment by WenyXu** [comment]
PTAL @fengjiachun @MichaelScofield

File: src/meta-srv/src/ddl.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+            .procedure_manager
+            .submit(procedure_with_id)
+            .await
+            .context(error::SubmitProcedureSnafu)?;
+
+        watcher::wait(&mut watcher)
+            .await
+            .context(error::WaitProcedureSnafu)?;
+
+        let status = rx.try_recv().context(error::TryReceiveStatusSnafu)?;
```
Comment (MichaelScofield) [inline src/meta-srv/src/ddl.rs:151]: Isn't `watcher::wait` not sufficient to indicate the create table procedure is done?

File: src/meta-srv/src/procedure/create_table.rs [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+}
+
+#[async_trait]
+impl Procedure for CreateTableProcedure {
+    fn type_name(&self) -> &str {
+        Self::TYPE_NAME
+    }
+
+    async fn execute(&mut self, _ctx: &ProcedureContext) -> ProcedureResult<Status> {
+        todo!()
```
Comment (MichaelScofield) [inline src/meta-srv/src/procedure/create_table.rs:223]: Can you implement it in this PR? I feel it's simply a match dispatch.

File: src/meta-srv/src/metasrv.rs [hunk 3]
```diff
@@ -109,6 +110,20 @@ pub struct SelectorContext {
     pub table: Option<String>,
 }
 
+impl SelectorContext {
+    pub fn from_ctx(catalog: String, schema: String, table: String, ctx: SelectorContext) -> Self {
```
Comment (MichaelScofield) [inline src/meta-srv/src/metasrv.rs:114]: Seems we can just clone the `SelectorContext`?

**Review by MichaelScofield — COMMENTED** [review]
Can't find any tests, are they in the following PRs?

**Comment by WenyXu** [comment]
> Can't find any tests, are they in the following PRs?

Yep, I will add tests in the following tests.

File: src/meta-srv/src/ddl.rs (hunk 1 above)
Comment (WenyXu) [reply src/meta-srv/src/ddl.rs:151]: We need a channel to receive the response from the procedure.(e.g., created table_id, etc. )

File: src/meta-srv/src/procedure/create_table.rs [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+            context,
+            creator: TableCreator { data },
+        })
+    }
+
+    fn global_table_key(&self) -> TableGlobalKey {
+        let table_ref = self.creator.data.table_ref();
+
+        TableGlobalKey {
+            catalog_name: table_ref.catalog.to_uppercase(),
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/meta-srv/src/procedure/create_table.rs:78]: Good catch. It's a typo😅

File: src/meta-srv/src/procedure/create_table.rs [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+        let mut joins = Vec::with_capacity(leaders.len());
+
+        for datanode in table_route.find_leaders() {
+            let client = clients.get_client(&datanode).await;
+            let client = Database::new(&table_name.catalog_name, &table_name.schema_name, client);
+
+            let regions = table_route.find_leader_regions(&datanode);
+            let mut create_expr_for_region = self.creator.data.task.create_table.clone();
+            create_expr_for_region.region_numbers = regions;
+            create_expr_for_region.create_if_not_exists = true;
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/meta-srv/src/procedure/create_table.rs:178]: > why explicitly set it true here? If you want to make `on_datanode_create_table` retryable, better check datanode's return value or status code.

You are right. To set it true is a shortcut to making an idempotent step. I will consider checking the status of the responses.

**Comment by WenyXu** [comment]
PTAL @MichaelScofield

File: src/meta-srv/src/procedure/create_table.rs [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+        match get_table_global_value(kv_store, key).await? {
+            Some(table_global_value) => {
+                // The metasrv crashed after metadata was created immediately.
+                // Recovers table_route from kv.
+                let table_id = table_global_value.table_id() as u64;
+                // If altering table operation (or other operations) doesn't hold the old table name as key-lock.
+                let expected = self.creator.data.table_route.table.id;
+                ensure!(
+                    table_id == expected,
+                    error::TableIdChangedSnafu {
```
Comment (MichaelScofield) [inline src/meta-srv/src/procedure/create_table.rs:178]: Under what circumstance could this table id be changed? Now the table metadatas are created atomically, other procedures should not update the table global value.

File: src/meta-srv/src/procedure/create_table.rs [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+    async fn on_datanode_create_table(&mut self) -> Result<Status> {
+        //FIXME: if datanode returns `Not Leader` error.
+        let table_route = &self.creator.data.table_route;
+
+        let table_name = self.table_name();
+        let clients = self.context.datanode_clients.clone();
+        let leaders = table_route.find_leaders();
+        let mut joins = Vec::with_capacity(leaders.len());
+
+        for datanode in table_route.find_leaders() {
```
Comment (MichaelScofield) [inline src/meta-srv/src/procedure/create_table.rs:204]:
```suggestion
        for datanode in leaders {
```

File: src/meta-srv/src/procedure/create_table.rs [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+            }
+        }
+
+        self.creator.data.state = CreateTableState::DatanodeCreateTable;
+
+        Ok(Status::executing(true))
+    }
+
+    async fn on_datanode_create_table(&mut self) -> Result<Status> {
+        //FIXME: if datanode returns `Not Leader` error.
```
Comment (MichaelScofield) [inline src/meta-srv/src/procedure/create_table.rs:196]:
```suggestion
```

File: src/meta-srv/src/procedure/create_table.rs (hunk 6 above)
Comment (WenyXu) [reply src/meta-srv/src/procedure/create_table.rs:178]: I guess if there is something like: Create table A, Create table A(from another Fe, Somehow, Failed), Renames table A to B, Create table A(Recovered)

**Review by MichaelScofield — APPROVED** [review]

File: src/meta-srv/src/procedure/create_table.rs [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+                    }
+                );
+            }
+            None => {
+                // registers metadata
+                self.register_metadata().await?;
+            }
+        }
+
+        self.creator.data.state = CreateTableState::DatanodeCreateTable;
```
Comment (fengjiachun) [inline src/meta-srv/src/procedure/create_table.rs:193]: What happens if a user makes a request when the metadata and routing information are complete, but the table creation is not yet finished?

File: src/meta-srv/src/service/ddl.rs [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+        key: id.to_string().into(),
+        table_id: Some(TableId {
+            id: table_id as u32,
+        }),
+        ..Default::default()
+    })
+}
+
+/// pre-calculates create table task's metadata.
+async fn handle_create_table_task_table_route(
```
Comment (fengjiachun) [inline src/meta-srv/src/service/ddl.rs:131]: The name is too long

Comment (fengjiachun) [reply src/meta-srv/src/service/ddl.rs:131]: handle_create_table_route ?

File: src/meta-srv/src/procedure/create_table.rs (hunk 9 above)
Comment (WenyXu) [reply src/meta-srv/src/procedure/create_table.rs]: I moved the `CreateMetadata` step after the `DatanodeCreateTable`.

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**